`f16` and `f128` step 3: compiler support & feature gate

tgross35 opened the pull request:
Continuation of https://github.com/rust-lang/rust/pull/121841, another portion of https://github.com/rust-lang/rust/pull/114607

This PR exposes the new types to the world and adds a feature gate. Marking this as a draft because I need some feedback on where I did the feature gate check. It also does not yet catch type via suffixed literals (so the feature gate test will fail, probably some others too because I haven't belssed).

If there is a better place to check all types after resolution, I can do that. If not, I figure maybe I can add a second gate location in AST when it checks numeric suffixes.

Unfortunately I still don't think there is much testing to be done for correctness (codegen tests or parsed value checks) until we have basic library support. I think that will be the next step.

Tracking issue: https://github.com/rust-lang/rust/issues/116909

r? @compiler-errors
cc @Nilstrieb
@rustbot label +F-f16_and_f128

tgross35 committed:
Add compiler support for parsing `f16` and `f128`

tgross35 committed:
Enable `f16` and `f128` in HIR

tgross35 committed:
Add feature gates for `f16` and `f128`

Includes related tests and documentation pages.

Michael Goulet: Don't issue feature error in resolver for f16/f128
unless finalize

Co-authored-by: Michael Goulet <michael@errs.io>

tgross35 committed:
Remove unneeded `f16` and `f128` parser tests

Superceded by feature gate tests.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8127381174/job/22212151238) [(plain)](https://github.com/rust-lang/rust/commit/a7ce0b8fc666b64506cb7677706b3cfd9cf87cf6/checks/22212151238/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=tgross35
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_a0e476ad-15a3-43d4-b7f9-a8c2d7e60112
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=f16-f128-step3-feature-gate
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_a0e476ad-15a3-43d4-b7f9-a8c2d7e60112
GITHUB_REF=refs/pull/121926/merge
GITHUB_REF_NAME=121926/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=789f321a94d1a6a5d6a3f570f25702905f6f8824
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_a0e476ad-15a3-43d4-b7f9-a8c2d7e60112
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_a0e476ad-15a3-43d4-b7f9-a8c2d7e60112
GITHUB_TRIGGERING_ACTOR=tgross35
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/121926/merge
GITHUB_WORKFLOW_SHA=789f321a94d1a6a5d6a3f570f25702905f6f8824
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_20_X64=/opt/hostedtoolcache/go/1.20.14/x64
---
#12 writing image sha256:96a851d0207a2695032930d5612123bf2bf105e1121d9d42ec7e869c6d345089 done
#12 naming to docker.io/library/rust-ci done
#12 DONE 9.9s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Sun Mar  3 03:11:26 UTC 2024
  network time: Sun, 03 Mar 2024 03:11:26 GMT
  network time: Sun, 03 Mar 2024 03:11:26 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'build.optimized-compiler-builtins', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
failures:

---- [ui] tests/ui/feature-gates/feature-gate-f128.rs stdout ----

error: /checkout/tests/ui/feature-gates/feature-gate-f128.rs:9: expected error not found: the feature `f128` is unstable
error: 0 unexpected errors found, 1 expected errors not found
status: exit status: 1
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/feature-gates/feature-gate-f128.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/feature-gates/feature-gate-f128" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/feature-gates/feature-gate-f128/auxiliary"
    Error {
        line_num: 9,
        kind: Some(
            Error,
---
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- [ui] tests/ui/feature-gates/feature-gate-f16.rs stdout ----

error: /checkout/tests/ui/feature-gates/feature-gate-f16.rs:9: expected error not found: the feature `f16` is unstable
error: 0 unexpected errors found, 1 expected errors not found
status: exit status: 1
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/feature-gates/feature-gate-f16.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/feature-gates/feature-gate-f16" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/feature-gates/feature-gate-f16/auxiliary"
    Error {
        line_num: 9,
        kind: Some(
            Error,
```

</details>

Nilstrieb commented on tests/ui/parser/f16-f128.rs:
```diff
@@ -1,37 +0,0 @@
-// Make sure we don't ICE while incrementally adding f16 and f128 support
-
-mod f16_checks {
-    const A: f16 = 10.0; //~ ERROR cannot find type `f16` in this scope
```
Your code handles this correctly, but can you add a new pass test with a custom struct called f16?

tgross35 replied:
This comment is in the sanity check from https://github.com/rust-lang/rust/pull/121728#discussion_r1505689228 which I removed here, but I do have that as part of the feature gate test https://github.com/rust-lang/rust/pull/121926/files/a7ce0b8fc666b64506cb7677706b3cfd9cf87cf6#diff-17cba0705a57a125f2d1eb3ddd12cee1ab97546c322cced8fcedec40a370f1a5R21-R34. Is there a better place for that?

tgross35 replied:
I put these in a different file so they'll stick around after the feature gate, https://github.com/rust-lang/rust/pull/121926/files#diff-139f82db73c1edb30cdb6cd881806233774bad18de0462af6b1e8bf1cef019e7

tgross35 committed:
Add UI tests related to feature-gated primitives

Add a test that `f16` and `f128` are usable with the feature gate
enabled, as well as a test that user types with the same name as
primitives are not improperly gated.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8135938638/job/22231342016) [(plain)](https://github.com/rust-lang/rust/commit/603aeb6c321cb3201cc8d6e7467d143859ce1d3a/checks/22231342016/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=tgross35
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_3f29fea8-954c-47be-a51c-1d6d3431cef1
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=f16-f128-step3-feature-gate
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_3f29fea8-954c-47be-a51c-1d6d3431cef1
GITHUB_REF=refs/pull/121926/merge
GITHUB_REF_NAME=121926/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=340483a85a40e87a0db2bcbf3ad9049a6f15322f
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_3f29fea8-954c-47be-a51c-1d6d3431cef1
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_3f29fea8-954c-47be-a51c-1d6d3431cef1
GITHUB_TRIGGERING_ACTOR=tgross35
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/121926/merge
GITHUB_WORKFLOW_SHA=340483a85a40e87a0db2bcbf3ad9049a6f15322f
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_20_X64=/opt/hostedtoolcache/go/1.20.14/x64
---
#12 writing image sha256:c75ad0bb2b3367a875f110dc49525bf495230a3fb5b6f662c451d2a9a2891469 done
#12 naming to docker.io/library/rust-ci done
#12 DONE 9.9s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Mon Mar  4 05:02:14 UTC 2024
  network time: Mon, 04 Mar 2024 05:02:14 GMT
  network time: Mon, 04 Mar 2024 05:02:14 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'build.optimized-compiler-builtins', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
+    |
+ LL | pub fn main() {
+    | ------------- previous definition of the value `main` here
+ ...
+ LL | fn main() {}
+    | ^^^^^^^^^ `main` redefined here
+    |
+    = note: `main` must be defined only once in the value namespace of this module
1 error[E0658]: the feature `f128` is unstable
-   --> $DIR/feature-gate-f128.rs:5:14
+   --> $DIR/feature-gate-f128.rs:3:10
3    |
3    |
- LL |     const A: f128 = 10.0;
-    |              ^^^^
+ LL | const A: f128 = 10.0;
6    |
7    = note: see issue #116909 <https://github.com/rust-lang/rust/issues/116909> for more information
8    = help: add `#![feature(f128)]` to the crate attributes to enable


9    = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date
11 error[E0658]: the feature `f128` is unstable
-   --> $DIR/feature-gate-f128.rs:8:16
+   --> $DIR/feature-gate-f128.rs:6:12
13    |
13    |
- LL |         let a: f128 = 100.0;
-    |                ^^^^
+ LL |     let a: f128 = 100.0;
16    |
17    = note: see issue #116909 <https://github.com/rust-lang/rust/issues/116909> for more information
18    = help: add `#![feature(f128)]` to the crate attributes to enable


19    = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date
21 error[E0658]: the feature `f128` is unstable
-   --> $DIR/feature-gate-f128.rs:13:15
+   --> $DIR/feature-gate-f128.rs:11:11
23    |
23    |
- LL |     fn foo(a: f128) {}
-    |               ^^^^
+ LL | fn foo(a: f128) {}
26    |
27    = note: see issue #116909 <https://github.com/rust-lang/rust/issues/116909> for more information
28    = help: add `#![feature(f128)]` to the crate attributes to enable


29    = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date
31 error[E0658]: the feature `f128` is unstable
-   --> $DIR/feature-gate-f128.rs:16:12
+   --> $DIR/feature-gate-f128.rs:14:8
33    |
33    |
- LL |         a: f128,
-    |            ^^^^
+ LL |     a: f128,
36    |
37    = note: see issue #116909 <https://github.com/rust-lang/rust/issues/116909> for more information
38    = help: add `#![feature(f128)]` to the crate attributes to enable


39    = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date
41 error[E0658]: the feature `f128` is unstable
-   --> $DIR/feature-gate-f128.rs:9:17
+   --> $DIR/feature-gate-f128.rs:7:13
43    |
---
46    |
47    = note: see issue #116909 <https://github.com/rust-lang/rust/issues/116909> for more information
48    = help: add `#![feature(f128)]` to the crate attributes to enable

49    = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date
- error: aborting due to 5 previous errors
+ error: aborting due to 6 previous errors
52 
- For more information about this error, try `rustc --explain E0658`.
- For more information about this error, try `rustc --explain E0658`.
+ Some errors have detailed explanations: E0428, E0658.
+ For more information about an error, try `rustc --explain E0428`.
54 


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/feature-gates/feature-gate-f128/feature-gate-f128.stderr
To only update this specific test, also pass `--test-args feature-gates/feature-gate-f128.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/feature-gates/feature-gate-f128.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/feature-gates/feature-gate-f128" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/feature-gates/feature-gate-f128/auxiliary"
--- stderr -------------------------------
error[E0428]: the name `main` is defined multiple times
##[error]  --> /checkout/tests/ui/feature-gates/feature-gate-f128.rs:17:1
   |
   |
LL | pub fn main() {
   | ------------- previous definition of the value `main` here
...
LL | fn main() {}
   | ^^^^^^^^^ `main` redefined here
   |
   = note: `main` must be defined only once in the value namespace of this module
error[E0658]: the feature `f128` is unstable
##[error]  --> /checkout/tests/ui/feature-gates/feature-gate-f128.rs:3:10
   |
   |
LL | const A: f128 = 10.0; //~ ERROR the feature `f128` is unstable
   |
   = note: see issue #116909 <https://github.com/rust-lang/rust/issues/116909> for more information
   = help: add `#![feature(f128)]` to the crate attributes to enable
   = help: add `#![feature(f128)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date
error[E0658]: the feature `f128` is unstable
##[error]  --> /checkout/tests/ui/feature-gates/feature-gate-f128.rs:6:12
   |
   |
LL |     let a: f128 = 100.0; //~ ERROR the feature `f128` is unstable
   |
   = note: see issue #116909 <https://github.com/rust-lang/rust/issues/116909> for more information
   = help: add `#![feature(f128)]` to the crate attributes to enable
   = help: add `#![feature(f128)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date
error[E0658]: the feature `f128` is unstable
##[error]  --> /checkout/tests/ui/feature-gates/feature-gate-f128.rs:11:11
   |
   |
LL | fn foo(a: f128) {} //~ ERROR the feature `f128` is unstable
   |
   = note: see issue #116909 <https://github.com/rust-lang/rust/issues/116909> for more information
   = help: add `#![feature(f128)]` to the crate attributes to enable
   = help: add `#![feature(f128)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date
error[E0658]: the feature `f128` is unstable
##[error]  --> /checkout/tests/ui/feature-gates/feature-gate-f128.rs:14:8
   |
   |
LL |     a: f128, //~ ERROR the feature `f128` is unstable
   |
   = note: see issue #116909 <https://github.com/rust-lang/rust/issues/116909> for more information
   = help: add `#![feature(f128)]` to the crate attributes to enable
   = help: add `#![feature(f128)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date
error[E0658]: the feature `f128` is unstable
##[error]  --> /checkout/tests/ui/feature-gates/feature-gate-f128.rs:7:13
   |
   |
LL |     let b = 0.0f128; //~ ERROR the feature `f128` is unstable
   |
   = note: see issue #116909 <https://github.com/rust-lang/rust/issues/116909> for more information
   = help: add `#![feature(f128)]` to the crate attributes to enable
   = help: add `#![feature(f128)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date
error: aborting due to 6 previous errors

Some errors have detailed explanations: E0428, E0658.
For more information about an error, try `rustc --explain E0428`.
```

</details>

compiler-errors commented on compiler/rustc_resolve/src/late.rs:
```diff
@@ -4123,6 +4126,7 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
                     && PrimTy::from_name(path[0].ident.name).is_some() =>
             {
                 let prim = PrimTy::from_name(path[0].ident.name).unwrap();
+                self.check_prim_gate(prim, path[0].ident.span);
```
Please inline this now since it's only called once.

tgross35 replied:
Done

compiler-errors reviewed (commented):
Are there any positive tests (i.e. ones that show that `feature(f16)`/`feature(f128)` actually allows us to name and use `f16`/`f128`)?

tgross35 commented on compiler/rustc_resolve/src/late.rs:
```diff
@@ -3923,7 +3924,9 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
                 }
 
-                if PrimTy::from_name(path[0].ident.name).is_some() {
+                if let Some(prim) = PrimTy::from_name(path[0].ident.name) {
```
Great, removed this

tgross35 commented on compiler/rustc_ast_passes/src/feature_gate.rs:
```diff
@@ -383,6 +383,17 @@ impl<'a> Visitor<'a> for PostExpansionVisitor<'a> {
+            ast::ExprKind::Lit(token::Lit { kind: token::LitKind::Float, suffix, .. }) => {
+                match suffix {
+                    Some(sym::f16) => {
+                        gate!(&self, f16, e.span, "the feature `f16` is unstable")
```
Changed

tgross35 commented on compiler/rustc_resolve/src/ident.rs:
```diff
@@ -421,6 +423,9 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
         let mut determinacy = Determinacy::Determined;
+        let sess = self.tcx.sess;
+        let feat_f16 = self.tcx.features().f16;
+        let feat_f128 = self.tcx.features().f128;
```
That was just to make the borrow checker happy, I now realize the closure gets `this` so it isn't needed 🤦

tgross35 commented on compiler/rustc_resolve/src/ident.rs:
```diff
@@ -598,7 +603,34 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
+                        Some(binding) => {
+                            match ident.name {
+                                sym::f16 => {
+                                    if !feat_f16 && !ident.span.allows_unstable(sym::f16) {
```
Done

tgross35 commented:
> Are there any positive tests (i.e. ones that show that `feature(f16)`/`feature(f128)` actually allows us to name and use `f16`/`f128`)?

I was thinking the feature gate test happened both with and without the gate but I guess not. Added the two `tests/ui/resolve/primitive-*` tests

RalfJung commented on tests/ui/resolve/primitive-f128.rs:
Do these have to be separate tests? There's overhead with each rustc invocation so IMO merging tests into a single file where it makes sense is generally preferable.

tgross35 replied:
As opposed to merging `f16` and `f128` into one? I can do that

tgross35 replied:
All done

compiler-errors commented:
@bors r+

bors commented:
:pushpin: Commit 1bc946b29041f7f2f0b77b9202d6a52fa79ca817 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 1bc946b29041f7f2f0b77b9202d6a52fa79ca817 -->
<!-- homu: {"type":"Approved","sha":"1bc946b29041f7f2f0b77b9202d6a52fa79ca817","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

RalfJung commented on compiler/rustc_middle/src/mir/interpret/value.rs:
```diff
@@ -211,6 +216,11 @@ impl<Prov> Scalar<Prov> {
+    #[inline]
+    pub fn from_f128(f: Quad) -> Self {
+        Scalar::Int(f.into())
+    }
```
There's `to_{f32,f64}` at the bottom of this file, we should probably have that for the new ones as well. (Can be a future PR though.)

RalfJung commented on compiler/rustc_middle/src/ty/consts/int.rs:
Similarly, this file has `try_to_f{32,64}` and should get the new ones.

tgross35 commented:
Thanks @RalfJung, I also figured the test might as well be run-pass so I just did those changes now.

RalfJung commented:
> I think that's the last of the compiler side for a while

I don't know your plans here -- the interpreter support is still mostly missing, right? This PR adds baic support for data of this type, but binops and casts with these types are still missing.

tgross35 commented:
> > I think that's the last of the compiler side for a while
> 
> I don't know your plans here -- the interpreter support is still mostly missing, right? This PR adds baic support for data of this type, but binops and casts with these types are still missing.

Maybe "a while" was an overstatement - I was planning to do basic library support next so the intrinsics are usable via traits (and I can finally reasonably test codegen), then circle back to anywhere that needed these to fill in or test against, which I think includes interpret.

Open to suggestions if this doesn't seem like the best path, I'm just loosely following https://github.com/rust-lang/rust/issues/116909#issuecomment-1969554030

compiler-errors commented:
@bors r+

bors commented:
:pushpin: Commit 9e00184f5570135eed3d93b7823cc4d99c9dd100 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 9e00184f5570135eed3d93b7823cc4d99c9dd100 -->
<!-- homu: {"type":"Approved","sha":"9e00184f5570135eed3d93b7823cc4d99c9dd100","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #121780) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

tgross35 commented:
One conflict in `ident.rs` resolved

@rustbot ready

oli-obk commented:
@bors r=compiler-errors

bors commented:
:pushpin: Commit b8f45a3189824a85750537d623820b35d9180399 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors b8f45a3189824a85750537d623820b35d9180399 -->
<!-- homu: {"type":"Approved","sha":"b8f45a3189824a85750537d623820b35d9180399","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

tgross35 commented:
@bors r-

Unfortunately it looks like this gating needs some work still https://github.com/rust-lang/rust/pull/122068#issuecomment-1980465791

tgross35 commented on compiler/rustc_resolve/src/ident.rs:
```diff
@@ -598,7 +600,36 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
-                        Some(binding) => Ok((*binding, Flags::empty())),
+                        Some(binding) => {
+                            match ident.name {
+                                sym::f16
+                                    if !this.tcx.features().f16
+                                        && !ident.span.allows_unstable(sym::f16) =>
+                                {
+                                    feature_err(
+                                        this.tcx.sess,
+                                        sym::f16,
+                                        ident.span,
+                                        "the type `f16` is unstable",
+                                    )
+                                    .emit();
+                                }
+                                sym::f128
+                                    if !this.tcx.features().f128
+                                        && !ident.span.allows_unstable(sym::f128) =>
+                                {
+                                    feature_err(
+                                        this.tcx.sess,
+                                        sym::f128,
+                                        ident.span,
+                                        "the type `f128` is unstable",
+                                    )
+                                    .emit();
+                                }
```
This is where the error is coming from but I am not sure why. In `tests/ui/resolve/conflicting-primitive-names.rs` it seems like the same-name types all get resolved before `Scope::BuiltinTypes` even gets checked. Not sure what exactly makes the `half` crate act differently than the test

compiler-errors replied:
Oh you know what? Try checking if `force` is true, and only issuing a feature error then. Otherwise, return `Err(Determinacy::Undetermined)`.

compiler-errors replied:
cc @petrochenkov who also may know better.

tgross35 replied:
Hm, returning an `Err(Determinacy::Undetermined)` from that block (in place of the `Ok((*binding, Flags::empty()))`) hits an ice at `compiler/rustc_resolve/src/lib.rs:2064`, need to dig a bit more.

What does `force` do here? I'm not sure how to replicate this in a unit test

tgross35 replied:
Well changing https://github.com/rust-lang/rust/blob/b8f45a3189824a85750537d623820b35d9180399/compiler/rustc_resolve/src/lib.rs#L2064 to `None` makes it compile with no ICEs and no FP on the gate lol. I need to trace back why that branch is expected to never hit

compiler-errors replied:
Actually, `Err(Determinacy::Determined)` "seems" right here. Unfortunately, literally none of this is documented as far as I can tell.

compiler-errors commented:
I believe I've fixed the resolver-based feature validation by checking `finalize.is_some()` before issuing an error here -- @petrochenkov, would you please offer some resolver knowledge to determine if this is the correct fix? Or if not, could you please let us know how to properly feature gate `f16`/`f128` here?

r? @petrochenkov

tgross35 commented:
(fyi I squished your fix into https://github.com/rust-lang/rust/pull/121926/commits/6eed834e09ff78c9170c814908ebd15d8861fb49)

petrochenkov commented:
It's probably simpler to gate this during AST -> HIR lowering.
When `hir::{Path,PathSegment}` is constructed, you need to put a `Res` into it, and if that `Res` is `Res::PrimTy(f16/128)` you can report a feature error.

But I'm not sure what exactly you want to gate.
- Does having a value of type `f16` (as opposed to type) unstable?
  If it's unstable then we either need to make sure that standard library doesn't return them from stable APIs (which is fine), or check types of all expressions/patterns/etc after type checking.
- Is using a type alias to `f16` unstable?
  If it's unstable then we either need to make sure that standard library doesn't create such stable aliases (which is fine), or check types of all HIR types after type checking.

If the standard library is "well behaving", then it should be enough to just check `Res`olutions in created HIR.
@rustbot author

tgross35 commented:
@petrochenkov  thanks for taking a look.

> It's probably simpler to gate this during AST -> HIR lowering. When `hir::{Path,PathSegment}` is constructed, you need to put a `Res` into it, and if that `Res` is `Res::PrimTy(f16/128)` you can report a feature error.

I tried something like this at some point, do you mean along the lines of what I have here? https://github.com/rust-lang/rust/commit/95b7f8a0908f28c0613911ceada6470b7b9e1f4b#diff-b24a095026827d389d25a6913f9426e6cea3c0ef9b348a12a834d3a5919b8687

> But I'm not sure what exactly you want to gate.
> 
>     * Does having a value of type `f16` (as opposed to type) unstable?
>       If it's unstable then we either need to make sure that standard library doesn't return them from stable APIs (which is fine), or check types of all expressions/patterns/etc after type checking.

I think we are OK if we don't throw the gate for something like `let val = othercrate::fn_returning_f16()`. Ideally we would flag this too, but it seems like diminishing returns because the other crate will already be forcing nightly and use of the feature gate. The standard library won't have any stable APIs returning `f16`/`f128` before the types themselves are stable.

>     * Is using a type alias to `f16` unstable?
>       If it's unstable then we either need to make sure that standard library doesn't create such stable aliases (which is fine), or check types of all HIR types after type checking.

I don't think there is any need to gate aliases, as long as it doesn't allow workarounds like the below:

```rust
type DefinitelyNotF16 = f16;

fn foo(a: DefinitelyNotF16) {}
```

Which I should probably add a test for.

> If the standard library is "well behaving", then it should be enough to just check `Res`olutions in created HIR. @rustbot author

Could you point me in the right direction here if there is anything different needed from what I linked above?

Some discussion where I was asking about this on Zulip, for context https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Where.20to.20gate.20a.20primitive

tgross35 commented:
Another question since I'm not very in the know here - is the reason you would prefer gating in HIR (rather than resolution) just because of simplicity, or is there a correctness aspect? The current approach roughly came from what was done for 128-bit integers https://github.com/rust-lang/rust/pull/49101/files#diff-3cff50afc88a11e641a6851dd19471a7e6470f05dd0cddaf2271e5cdd9936b73.

petrochenkov commented:
@tgross35 
> Another question since I'm not very in the know here - is the reason you would prefer gating in HIR (rather than resolution) just because of simplicity, or is there a correctness aspect? The current approach roughly came from what was done for 128-bit integers https://github.com/rust-lang/rust/pull/49101/files#diff-3cff50afc88a11e641a6851dd19471a7e6470f05dd0cddaf2271e5cdd9936b73.

Just because of (possible) simplicity.
If the current approach worked for `u128` then it will work for `f128` as well.
*Unless* standard library reexports `f128` similarly to https://doc.rust-lang.org/stable/std/primitive/index.html#reexport.u128, then it will be available via a full path `std::primitive::f128` and not just in scope, and the current gating doesn't cover that case.

petrochenkov commented:
https://github.com/rust-lang/rust/commit/95b7f8a0908f28c0613911ceada6470b7b9e1f4b#diff-b24a095026827d389d25a6913f9426e6cea3c0ef9b348a12a834d3a5919b8687 is not enough, it covers types but not imports, for example (`use f128 as foo;`).
That's why I suggested doing it in AST lowering when `hir::{Path,PathSegment}` are constructed, then it will cover all paths.

Also, both the current implementation and https://github.com/rust-lang/rust/commit/95b7f8a0908f28c0613911ceada6470b7b9e1f4b#diff-b24a095026827d389d25a6913f9426e6cea3c0ef9b348a12a834d3a5919b8687 use feature gating incorrectly.
It doesn't need to return some `ty::Err` or `Err(Determinacy::Determined)` if feature gate fires, it just needs to emit an error, but then continue as usual and return f16/f128.

tgross35 commented:
@petrochenkov I changed this to no longer emit a determinacy error in resolution. Wouldn't doing the gate in `rustc_ast_lowering` rather than after resolution mean that user-defined types named `f16` or `f128` also get flagged? There are also 9 different places in lowering where a `hir::Path` is created, is there a more common point where these created paths could be checked?

I don't think accidentally exposing these types via reexports in `std` is much of a concern, since anyplace they are available in the libraries will have a feature gate there.

petrochenkov commented:
> Wouldn't doing the gate in `rustc_ast_lowering` rather than after resolution mean that user-defined types named `f16` or `f128` also get flagged?

User defined types will result in `Res::Def`, not `Res::PrimTy`.

>I don't think accidentally exposing these types via reexports in `std` is much of a concern, since anyplace they are available in the libraries will have a feature gate there.

Stability attributes on reexports do not work, so it's still a concern.
We just need to explicitly avoid exposing these types through the library, until they are stable.

petrochenkov commented:
@bors r=compiler-errors,petrochenkov

bors commented:
:pushpin: Commit 2098fec0809521ea8dd489f6cd5f09337a31764f has been approved by `compiler-errors,petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors,petrochenkov 2098fec0809521ea8dd489f6cd5f09337a31764f -->
<!-- homu: {"type":"Approved","sha":"2098fec0809521ea8dd489f6cd5f09337a31764f","approver":"compiler-errors,petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

tgross35 commented:
> > Wouldn't doing the gate in `rustc_ast_lowering` rather than after resolution mean that user-defined types named `f16` or `f128` also get flagged?
> 
> User defined types will result in `Res::Def`, not `Res::PrimTy`.

I only see `Res::PrimTy` being constructed once in ast_lowering at https://github.com/rust-lang/rust/blob/72d78970ec7cb5cdf210290835d9f71d4d6663e2/compiler/rustc_ast_lowering/src/item.rs#L1580-L1593, which looks to be for generics. The rest are in `rustc_resolve/src/late.rs`, which we intercept here.

Thanks for the r+. As-is should be workable for now then, I can later update this gating to your suggestion if it improves the situation with libs. I am just not exactly sure what this would look like still.

matthiaskrgr commented:
@bors rollup=iffy

bors commented:
:hourglass: Testing commit 2098fec0809521ea8dd489f6cd5f09337a31764f with merge 76a9141eaa71e023ba959da0066bf4da5e6bf287...
<!-- homu: {"type":"BuildStarted","head_sha":"2098fec0809521ea8dd489f6cd5f09337a31764f","merge_sha":"76a9141eaa71e023ba959da0066bf4da5e6bf287"} -->

rust-log-analyzer commented:
The job **`aarch64-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/8303435418/job/22727543229) [(plain)](https://github.com/rust-lang-ci/rust/commit/76a9141eaa71e023ba959da0066bf4da5e6bf287/checks/22727543229/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] rustc_infer test:false 92.838
   Compiling rustc_mir_build v0.0.0 (/checkout/compiler/rustc_mir_build)
[RUSTC-TIMING] rustc_codegen_ssa test:false 90.945
   Compiling rustc_lint v0.0.0 (/checkout/compiler/rustc_lint)
##[error]The runner has received a shutdown signal. This can happen when the runner service is stopped, or a manually started runner is canceled.
##[group]Clock drift check
  local time: Sat Mar 16 00:01:55 UTC 2024
  local time: Sat Mar 16 00:01:55 UTC 2024
Session terminated, killing shell... ...killed.
##[error]The operation was canceled.
Cleaning up orphan processes
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8303435418/job/22728162850)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/8303435418/job/22728162850","builder_name":"checks-actions"} -->

tgross35 commented:
Clock drift failed, I don’t think that’s from the changes

compiler-errors commented:
@bors retry

bors commented:
:hourglass: Testing commit 2098fec0809521ea8dd489f6cd5f09337a31764f with merge c03ea3dfd907e7dc6305ebf20c94f3a1f1d9fed7...
<!-- homu: {"type":"BuildStarted","head_sha":"2098fec0809521ea8dd489f6cd5f09337a31764f","merge_sha":"c03ea3dfd907e7dc6305ebf20c94f3a1f1d9fed7"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8304546361/job/22731792053)
Approved by: compiler-errors,petrochenkov
Pushing c03ea3dfd907e7dc6305ebf20c94f3a1f1d9fed7 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"compiler-errors,petrochenkov","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8304546361/job/22731792053"},"merge_sha":"c03ea3dfd907e7dc6305ebf20c94f3a1f1d9fed7"} -->

rust-timer commented:
Finished benchmarking commit (c03ea3dfd907e7dc6305ebf20c94f3a1f1d9fed7): [comparison URL](https://perf.rust-lang.org/compare.html?start=05a2be3def211255dc7640b006ac10f0f02baf5c&end=c03ea3dfd907e7dc6305ebf20c94f3a1f1d9fed7&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=05a2be3def211255dc7640b006ac10f0f02baf5c&end=c03ea3dfd907e7dc6305ebf20c94f3a1f1d9fed7&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=05a2be3def211255dc7640b006ac10f0f02baf5c&end=c03ea3dfd907e7dc6305ebf20c94f3a1f1d9fed7&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=05a2be3def211255dc7640b006ac10f0f02baf5c&end=c03ea3dfd907e7dc6305ebf20c94f3a1f1d9fed7&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=05a2be3def211255dc7640b006ac10f0f02baf5c&end=c03ea3dfd907e7dc6305ebf20c94f3a1f1d9fed7&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 670.3s -> 670.952s (0.10%)
**Artifact size**: 311.50 MiB -> 311.63 MiB (0.04%)
